### ISSUE TITLE
Provide (optional) ability to use Kubernetes as the runtime engine instead of Docker

### ISSUE THREAD

**Issue by JeanMertz** [drone/drone#1815]
I was wondering, has there been a consideration to move drone up the stack, and have it rely on Kubernetes features to function?

It could potentially ease the burden of a lot of things Drone currently has to manage on top of Docker, and with minikube, running drone locally could become as simple as `minikube create && drone deploy`.

I know this is an extreme oversimplification of things, and it would obviously mean giving up some freedom (dependent on a container scheduler, instead of only Docker), but there are obviously also a lot of upsides to this.

**Comment by bradrydzewski** [comment]
@JeanMertz right now there are no plans for this sort of deep integration with kubernetes. I'm certainly keeping an eye on the project and where it goes in the future, but I'm not sure this would be the right decision for drone at this time.

**Comment by bradrydzewski** [comment]
I should point out however, that the engine used to run builds (docker) is exposed as an interface so in theory it could be swapped with a different implementation. This is the interface that is defined for running builds:
https://github.com/drone/drone/blob/master/build/engine.go

And this is the docker implementation:
https://github.com/drone/drone/tree/master/build/docker

I would certainly encourage a community effort to create a kubernetes implementation. I know @tboerger expressed some interest here. I've discussed with @gtaylor as well. Bottom line is I think making drone kubernetes-only would not be a wise decision for the project, but supporting multiple runtime engines is certainly of interest.

This sort of thing, of course, depends on community engagement since I cannot volunteer to take on all these tasks. So while it isn't something I would work on, I would certainly make myself available for individuals interested in contributing such functionality.

**Comment by JeanMertz** [comment]
That engine abstraction looks interesting.

Kubernetes has a very good first-class Golang API, and the feature set required here (Start/Stop/Remove/Wait/Logs) seems really limited, so it wouldn't be _too_ hard to implement that on top of Kubernetes.

Maybe I'll give it a stab some time in the near future, if I ever manage to put more than 24 hours in a day.

**Comment by bradrydzewski** [comment]
cool, if you end up looking into an implementation give a shout-out in the drone developer channel at https://gitter.im/drone/drone-dev . I'm sure you could find some others interesting in lending a hand :)

**Comment by bradrydzewski** [comment]
Let's re-open this but with a slightly adjusted scope of adding experimental support for an alternate kubernetes engine, alongside the existing docker engine. I would love to hear what @gtaylor thinks about this and what might be possible.

**Comment by tboerger** [comment]
When I'm more familiar with the k8s codebase I really would like to give it a try to build a real k8s agent. That's something my teamlead also asked for.

**Comment by gtaylor** [comment]
I've got a crappy custom scheduler that can parse a .drone.yml and fire up pods with some Drone plugins working. This isn't useful for Drone itself, but served as a nice exercise to get a feel for what this would look like. A few notes about what I did:

* Each build gets its own Kubernetes Namespace. This makes cleanup as easy as one DELETE call, and also means we don't need to get weird with service names (to avoid collisions). Eventually we'll be able to set network ACLs to prevent cross-Namespace interactions, and we can already set quotas to avoid resource exhaustion.
* Each build is a Job (not a Replication Controller or Deployment). Key differentiator is that we don't restart on failure.
* In my experimentation, all of the pipeline is in one Pod so we can mount the same emptyDir volume on all of the pipeline containers.
* Dependent services run in separate Pods. These are accessible by hostname.

Things I haven't got around to:

* .netrc injection.
* I'm not sure if multiple containers within one Pod start in order, and block on their predecessor. We wouldn't want multiple pipeline steps executing simultaneously by default. In an ideal case, we can run all pipeline steps in one Pod. In a less ideal case, we run each pipeline as a step and pass around a volume, which would potentially be much slower.
* As Brad mentioned, secrets. One way to address this is for Drone to do the figuring out what should be injected into the build's Namespace as a Secrets object. In other words: the Kubernetes part should be handed the secrets that have already been resolved.
* Build Namespaces should probably eventually be cleaned up. Drone can pull the logs and the build states as it goes, so the only reason to keep them around would be for troubleshooting. This is going to vary by org, but we'd probably want to keep them for 24 hours over here.

**Comment by bradrydzewski** [comment]
@gtaylor thanks for the detailed reply. Regarding using kubernetes secret store, I'm wondering how well that would work with drone. Thoughts on https://github.com/drone/drone/issues/1808#issuecomment-252933474 ?

**Comment by gtaylor** [comment]
@bradrydzewski I think you create multiple Kubernetes Secret resources, perhaps one per pipeline step (a container within the build Pod). You can pull these secrets into each container individually. Drone would determine which secrets go to which steps (or services), and stuff those secrets in the step's respective Secret object.

Also note that each step (Pod and at least one container) can have multiple Secret objects to pull env vars (or mount as files) from. That may or may not be useful to you.

While what I am describing above doesn't lend itself to a ton of value over what Drone has, the values won't be viewable in `kubectl describe pod`, unlikey straight env vars. It'd be very important not to show secret values in pod descriptions.

**Comment by gtaylor** [comment]
Also, I don't think it'd be worth heavily pursuing deep Kubernetes integration until Kubernetes 1.5 lands. The Job system is still shaking out in major ways. Secrets are going to be seeing lots of expansion soon, as are the various network/resource ACLs. 

It'd certainly be worth tinkering with and building some familiarity, but this is going to take a good bit of thought and care to do well. We'd need to get kind of hacky with build pods and plugins to make it work right now.

**Comment by derekperkins** [comment]
@gtaylor Now that 1.5 is out, do you feel more confident about tackling this?

**Comment by bradrydzewski** [comment]
FWIW I am also interested, at some point, trying to figure out what a "serverless" drone would look like. I think the concept of a build queue and pending builds could be eliminated by using the on-demand capabilities of services like hyper.sh and rackspace carina. I'm sure other vendors will launch similar on-demand capabilities as well.

I'm not sure how Kubernetes fits into the picture here, but am interested in the overall concept.

**Comment by gtaylor** [comment]
@derekperkins It's definitely more possible now. It would still be a whole lot of work to do really well, in that the perfect situation is that we're scattering the work out across multiple Pods. Failing to achieve that means that we're not any better off than we currently are.

It's one of those things where this could be really awesome if done right, but it could also be thoroughly underwhelming and a black mark otherwise. We'd have to provide something more compelling and capable than all of these Jenkins + Kubernetes whitepapers (a well-trodden path at this point), at the very minimum.

> FWIW I am also interested, at some point, trying to figure out what a "serverless" drone would look like.

It could be neat, but is there any money in that? At what point do you just run Circle CI/Travis/CodeShip/Shippable or one of the infinite other hosted solutions that are effectively "serverless" from the customer's perspective? Can't imagine the bigger money on-prem orgs using those services with their metal.

If you really don't want to maintain servers, fire up a Google Container Engine (hosted Kubernetes) cluster and install Drone. They maintain the VMs and it's cheap ($5/month at the lowest level). You can still get your fingers in if you want to have the cluster auto-scale up/down as jobs pile up, and you can mix in their equivalent of Spot instances (pre-emptible VMs). If and when you eventually want to take more direct control with your own cluster, Container Engine runs the same Kubernetes that is found in the open source project's repo.

> I'm not sure how Kubernetes fits into the picture here

It's still probably a little early for Drone _and_ Kubernetes to go down this road too much yet, but it fits into the picture in that it's not a proprietary, close sourced option like hyper.sh and Rackspace Carina :) It almost assuredly now has far more adoption and mindshare than those two.

**Comment by bradrydzewski** [comment]
I'm going to list some challenges based on my conversation with @gtaylor. I'm not a kubernetes expect, so my apologies if I misinterpreted the discussion.

1. kubernetes has no notion of sequential or chained jobs. This will need to be simulated
2. drone uses a single shared network for all containers in the build process. This means service containers are available at localhost. This could pose a challenge depending on how we implement kubernetes support. Using multiple-pods would prevent this
3. drone support single-machine fan-in / fan-out. This means build steps can be running in parallel and have access to the underlying build workspace (where the code is cloned). In kubernetes, if using multiple-pods, this could be difficult since we can only mount the volume to a single pod at a time.

These are some of the main challenges that we will face with native kubernetes support, as I understand it.

**Comment by JeanMertz** [comment]
As a reference point: we are currently using Jenkins on top of Kubernetes, together with some plugins (one of them being the [kubernetes-plugin](https://github.com/jenkinsci/kubernetes-plugin)), to simulate what I'd like Drone to do/represent.

Jenkins comes with a lot of bagage (mostly good, some bad, some ugly), but the current set-up looks something like this:

- We run Jenkins on top of GKE (Google Container Engine)
- the concept of _build nodes_ is translated to single pods
  - each job on Jenkins creates a single pod
  - this pod represents a one-time-use "node"
  - within this pod, the job is started
  - we run our tests in parallel, for each parallel process, we launch one container, so some tests run 22 containers in a single pod
  - this ensures `localhost` works across containers
  - it also ensures data is shared across volumes
- we also have the "autoscale" feature of GKE enabled, this means that if too many jobs are being scheduled, GKE will start up a new node and add it to the node-pool of out Kubernetes cluster
- all our nodes run on preemptible machines (cheaper, but "unreliable"), in practice, this means once every 1000 or so runs, a job fails because the node was deleted on GCE, but we accept this
- in effect, this means we have near-infinite scalability of our CI environment

**Comment by bradrydzewski** [comment]
@JeanMertz is this something you would be willing to help implement?  I have no real world experience with Kubernetes and have quite a lot on my plate. Perhaps if this were a community effort it would have more of a chance of succeeding. What do you think?

**Comment by jmn** [comment]
Hi,
> kubernetes has no notion of sequential or chained jobs. This will need to be simulated

I am not sure if this is what is meant but there are [Init Containers](https://kubernetes.io/docs/user-guide/pods/init-container/):

> An init container is exactly like a regular container, except that it always runs to completion and each init container must complete successfully before the next one is started.

**Comment by bradrydzewski** [comment]
@jmn I think perhaps a better way of describing the issue is that kuberenetes does not easily map to the drone yaml at this time. The drone yaml executes batch steps, with linked services, and needs to evaluate whether or not the step should be executed at runtime based on results of prior steps.

Consider this configuration:

```
pipeline:
  backend:
    image: golang
    commands:
      - go build
      - go test
  frontent:
    image: node
    commands:
      - npm install
      - npm build
  publish:
    image: plugins/docker
    repo: foo/bar
    when:
      event: push
  deploy:
    image: plugins/ssh
    shell:
      - docker pull foo/bar
      - docker stop foo/bar
      - docker run foo/bar
    when:
      event: deployment
      branch: master
  notify:
    image: plugins/slack
    channel: dev
    when:
      status: [ success, failure ]

services:
  redis:
    image: redis:latest
  mysql:
    image: mysql:latest
```

This doesn't mean it is impossible, though. The suggestion by @JeanMertz is really interesting. His suggestion is that each step should be its own pod, with its own set of services, and Drone would handle orchestrating sequential pod execution to emulate build steps.

Unfortunately I do not have any experience with kubernetes outside of reading a few blog posts, so it is not something I will be able to implement at this time. Community contributions very welcome :)

**Comment by bradrydzewski** [comment]
I should point out that I'm also not connected to the kubernetes community. If there are individuals in the kubernetes community that you think might be interested in working on this feature, you should send them to the drone gitter channel.

**Comment by webwurst** [comment]
I would like to help out where I can. We are using a small/cheap Kubernetes
cluster for some time for some open-data projects:
https://github.com/codeformuenster/kubernetes-deployment

And we used Drone to create Docker images for ARM a while ago:
https://github.com/armhf-drone-plugins

Haven't played with Drone 0.5 yet unfortunately. And constraint would be
time, as always ;)

**Comment by derekperkins** [comment]
@clintberry just got it setup and is working on a `helm install` package

**Comment by benschumacher** [comment]
I've used a slightly modified version of this repo from @vallard to deploy Drone to Kubernetes:

https://github.com/vallard/drone-kubernetes

Works well enough, though it does bypass the Kubernetes scheduler, and connects to the docker daemon on the host directly. Would definitely be interested in a solution that could work within that context.

One note: I think trying to map some of Drone's concepts to Kubernetes directly isn't going to help make this happen. In general, I think the design @gtaylor suggested above has some merit, though I'm not sold that having a new namespace-per-build is necessary. Working within the context of a pod isn't too far off what Drone is doing right now to run builds, it would just require something Kubernetes-specific logic to ensure that various "components" within a Pod are a) started in the right order, and b) executed serially, c) etc. Keep in mind that the main benefit of a Pod within Kubernetes is to link together dependent containers w/o relying on cluster-wide functions like service discovery, services, ingress controllers, etc.

I think starting w/ a set of requirements around what is expected from build scheduler, including some of the newer features around shared volumes, matrix builds, fan-out, fan-in, could help clarify what is required. The effort should start small, too, just solve a simple use case that 1) clones a repo, 2) executes a build step, 3) collects logs from these steps. I'd be happy to carve out some time to look into this deeper, but I doubt I'd have much time to contribute much in the way of coding in the near future.

**Comment by bradrydzewski** [comment]
@benschumacher my ultimate goal is to create a compiler of sorts (note that it is not entirely vaporware, as I do have certain pieces working). The compiler would consist of frontends which would take different configuration formats (drone.yml, travis.yml, bitbucket-pipelines.yml) and compile them down to an intermediate representation. I have an implementation of this that works with the .drone.yml and bitbucket-pipelines.yml.

I am working on a formal specification for the intermediate representation:
https://github.com/cncd/cncd/blob/master/content/spec/ir.md

Ideally this intermediate representation would work with multiple backends, where a backend is a container engine (such as Docker) or an orchestration engine (such as Kubernetes). I have a working backend for Docker and am confident this could work for LXD and Rocket. I am not sure what changes would be required for this to work with Kubernetes, however, I feel like this could be a solvable problem.

>  think starting w/ a set of requirements around what is expected from build scheduler, including some of the newer features around shared volumes, matrix builds, fan-out, fan-in, could help clarify what is required

I think the specification for the IR gets us closer to this goal. It might still be too docker specific though, so would love to hear your feedback, and would love to have you as part of the working group (that invite extends to anyone in this thread as well).

> I'd be happy to carve out some time to look into this deeper, but I doubt I'd have much time to contribute much in the way of coding in the near future.

At this stage if you can participate in more of an architectural (non-coding) capacity it would still be tremendously helpful. The backend implementations tend to be quite small, so if we get the IR specification in a good place the implementation should hopefully be pretty straightforward.

**Comment by clintberry** [comment]
Forgive my ignorance, but I don't understand what you are all trying to accomplish here. Kubernetes just uses docker under the hood, and Drone runs great in kubernetes right now. I installed the drone server as a deployment, and drone agents as a deployment. I can scale the agents now with 2 clicks as needed. I don't think adding a tighter integration to kubernetes gives you anything special, but this is where my ignorance comes in. What features are you looking for with a deeper kubernetes integration?

**Comment by kop** [comment]
> Forgive my ignorance, but I don't understand what you are all trying to accomplish here. Kubernetes just uses docker under the hood, and Drone runs great in kubernetes right now. I installed the drone server as a deployment, and drone agents as a deployment. I can scale the agents now with 2 clicks as needed. I don't think adding a tighter integration to kubernetes gives you anything special, but this is where my ignorance comes in. What features are you looking for with a deeper kubernetes integration?

Such setup is very limited. With Kubernetes scheduler the following features can be archived:

- Avoid Docker In Docker usage;
- "Real" scaling, when K8S pod is created for every task and cluster scales automatically up and down to fulfil CPU/RAM requests;
- When you run DIND setup, any of services you create in your `drone.yml` will be deployed to the same host your agent is running in. This is bad, they should deploy to the least busy host instead;
- I believe there are a lot more, those are just cases that came to my head instantly.

**Comment by clintberry** [comment]
> Avoid Docker In Docker usage

I don't know if I am comfortable letting my CI engine spin up Kubernetes pods in production. I think I would rather keep the Docker-in-Docker methods for isolation/security. I'm sure you could still get some sort of isolation/security with drone-created pods, but is it worth the hassle? 

> "Real" scaling...

I am new to drone, so maybe I was wrong in assuming this, but I assumed I would be able to run only one concurrent job/build per agent that is connected to my drone server. For me, this is ideal because I can control the amount of resources that my build system uses. I don't want infinite scaling of my build system using precious production resources.

> When you run DIND setup, any of services you create in your drone.yml will be deployed to the same host your agent is running in. This is bad, they should deploy to the least busy host instead;

I can see why you would like that. Especially if you have large services you need to spin up. I still don't want anything running outside of my agent docker, but I totally could see why you would want this. But at the same time, each agent gets distributed to kube workers according to load, so you get at least some distribution of resources, but certainly not to the level you are suggesting here.

I understand I am probably being too narrow-minded on this. I apologize if I am coming across confrontational. I am just trying to understand your use cases a bit more.

**Comment by kop** [comment]
> I don't know if I am comfortable letting my CI engine spin up Kubernetes pods in production. I think I would rather keep the Docker-in-Docker methods for isolation/security. I'm sure you could still get some sort of isolation/security with drone-created pods, but is it worth the hassle?

I don't think it's the case really. K8S offers way better isolation than DIND.
Since the DIND instances run in privileged mode, they can easily escalate to the host, and you probably don’t want this.

> I am new to drone, so maybe I was wrong in assuming this, but I assumed I would be able to run only one concurrent job/build per agent that is connected to my drone server. For me, this is ideal because I can control the amount of resources that my build system uses. I don't want infinite scaling of my build system using precious production resources.

I don't really follow Drone development right now. And use Jenkins instead only because it has K8S integration.
How it works in Jenkins: build nodes are created dynamically when a build is queued. I can control the limit of how much agents I allow to run concurrently. When my build queue is empty - I run no nodes and don't pay for anything. When I have a bunch of builds to run - nodes are created automatically to allow fast thruput of my builds. For me this is ideal.

> I understand I am probably being too narrow-minded on this. I apologize if I am coming across confrontational. I am just trying to understand your use cases a bit more.

We are here to discuss and find best solutions, don't we? :)

**Comment by JeanMertz** [comment]
> I don't think it's the case really. K8S offers way better isolation than DIND.
> Since the DIND instances run in privileged mode, they can easily escalate to the host, and you probably don’t want this.

Indeed. It's _far_ safer to let Kubernetes manage the scheduling of the containers for you (since that's what it is, a container scheduler), than to give free-reign access to your Docker daemon without any checks and balances on how it will behave.

Furthermore, if you cut out the scheduler in this case, then Kubernetes no longer has any accurate information on freely available system resources, since it doesn't manage them anymore, and thus your system will become more unstable when going the DinD route.

**Comment by clintberry** [comment]
I thought it was running docker inside of my drone agent pod? My drone agent is controlled by the scheduler and k8s knows its resources. I mean, you do mount the folder for the docker runtime, but you mount it inside your container and it executes inside the container. Perhaps I am missing something technically here.

**Comment by JeanMertz** [comment]
@clintberry yes, apologies. You are correct. I was confusing the DinD setup with the shared docker daemon setup.

In your case, you run a _new_ Docker daemon inside a Docker container.

Still though, as @kop mentioned, you still need to run that initial (Kubernetes-managed) container in privileged mode, which increases the risk of problems.

Also, there's [this blogpost from the original Docker in Docker creator](http://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/) warning about using such a system specifically for CI usage.

**Comment by clintberry** [comment]
Okay, that all makes sense... I guess for now I will make a specific kube worker for CI and mount the host docker socket. But I see why you are all doing what you are doing now. Thanks for the explanations :-)

**Comment by bradrydzewski** [comment]
I can provide a bit more context. This is going to be a bit long winded, so I apologize in advance ...

We are spending a lot of time building a world-class scheduler for Drone. This scheduler needs to distribute builds across many servers, taking resource usage, architecture (windows vs linux) and user-defined parameters into account. The scheduler needs to be efficient, fault tolerant and stable. This piece of software is incredibly tedious to write, and it is easy to introduce bugs that are hard to reproduce and debug. The scheduler represents a large chunk of the Drone codebase, and takes up a significant amount of my time.

Kubernetes and Swarm and Hyper and ECS and others also have advanced scheduling and orchestration capabilities. They also have auto-scaling. There is a clear overlap in functionality that has me asking myself if I should continue to invest so much of my time in Drone's internal scheduling and orchestration capabilities, especially given the pace of development and number of man hours being invested in solutions like Kubernetes.

I think the smart move is for Drone to have a very basic embedded scheduler for small installations, but to use Kubernetes / Swarm / etc for larger scale and more elastic installations, with more advanced resource scheduling needs.

I also think we should really re-consider the build queue. Right now every CI system out there (Jenkins, Travis, Circle, Drone and others) has a build queue. You see builds sitting in a queue waiting for an available server. With high volume you could see your build sitting in the queue for 30 minute or more.

In a world where we have elastic compute and are pushing the idea of serverless, the build queue feels so 2003. I would like to remove the build queue from Drone and rely on orchestration systems like  Kubernetes / Swarm / etc to handle elasticity so that a build is never pending ever again. If a cluster lacks capacity, it is automatically increased to execute your build with little or no wait.

Imagine one day you download and configure Drone with access to your Kubernetes environment and you automatically have the ability to elastically scale from 1 to 1000 builds servers. Or you hook it up to [Hyper](http://hyper.sh/) with a seemingly infinite pool of bare metal servers that bill you by the second for running container instances, so that you no longer pay for idle servers and unused compute. This is the future I want for a CI system and the future I want for the Drone community.

**Comment by derekperkins** [comment]
> Imagine one day you download and configure Drone with access to your Kubernetes environment and you automatically have the ability to elastically scale from 1 to 1000 builds servers. Or you hook it up to Hyper with a seemingly infinite pool of bare metal servers and per-second billing, so that you no longer pay or idle servers and unused compute time. This is the future I want.

So much winning. :) I think it's a very smart move to outsource the scheduling to purpose built scheduling systems, and focus your dev time on true differentiators.

> I would like to remove the build queue from Drone and rely on orchestration systems like Kubernetes / Swarm / etc to handle elasticity so that a build is never pending ever again. If a cluster lacks capacity, it is automatically increased to execute your build with little or no wait.

I'm not sure that the queue should disappear. While waiting for builds is annoying, that will always be an option people will need/want. If you're running kubernetes on your own bare metal, you're not necessarily going to be able to auto-scale. Same goes if you're under strict budget restrictions. In those cases, you are going to be ok waiting until spare capacity is available.

For people who do have access to autoscaling however, drone could export metrics like "jobsInQueue", that you could use as a scaling metric for drone workers, making it possible to run as many in parallel as you want.

**Comment by gtaylor** [comment]
>  If you're running kubernetes on your own bare metal, you're not necessarily going to be able to auto-scale

This doesn't really matter, since you're firing off Kubernetes Jobs. You end up with queue-like behavior. Whether you choose to auto-scale it doesn't matter.

**Comment by clintberry** [comment]
> This doesn't really matter, since you're firing off Kubernetes Jobs. You end up with queue-like behavior. Whether you choose to auto-scale it doesn't matter.

Yes, if kubernetes doesn't have enough resources, it will wait until it does. But I may not want to max out my k8s cluster. I, for one, would like to put limits on how many resources my build process can use and queue up anything beyond that, without sucking up all the ram on my cluster.

**Comment by bradrydzewski** [comment]
> Yes, if kubernetes doesn't have enough resources, it will wait until it does. But I may not want to max out my k8s cluster.

I believe this is a problem that can be solved. Kubernetes has the ability to assign pods to a subset of nodes in your cluster. This should be conceptually no different than installing the agent on a subset of nodes in your cluster, thus limiting the nodes on which builds are executed.

... but I think we are getting ahead of ourselves. I am looking for help defining a specification and creating a __proof of concept__ to schedule builds using Kubernetes as a backend. Right now I'm most interested in proving whether or not it is possible. It is absolutely possible we may find a Kubernetes backend doesn't make sense.

So if and when we have a working proof of concept we can present to the community and collect feedback and potential concerns and edge cases.

**Comment by bradrydzewski** [comment]
> I agree with your vision. Drone is awesome and my company will help support you with whatever direction you take.

and thanks!  If you are interested in helping shape how Drone runs on Kubernetes I would definitely encourage you to join the (soon to be formed) working group.

**Comment by derekperkins** [comment]
> This doesn't really matter, since you're firing off Kubernetes Jobs. You end up with queue-like behavior. Whether you choose to auto-scale it doesn't matter.

For sure, that makes lots of sense. I wasn't sure how deeply integrated you were thinking.

> Yes, if kubernetes doesn't have enough resources, it will wait until it does. But I may not want to max out my k8s cluster. I, for one, would like to put limits on how many resources my build process can use and queue up anything beyond that, without sucking up all the ram on my cluster.

The best part about this proposal is that drone shouldn't have to worry about your resource consumption, but instead, rely on kubernetes to expose resource restrictions for jobs.

**Comment by gtaylor** [comment]
> But I may not want to max out my k8s cluster. I, for one, would like to put limits on how many resources my build process can use and queue up anything beyond that, without sucking up all the ram on my cluster.

Then put resource limits on your namespace. All of these concerns can be addressed with Kubernetes built-ins.

**Comment by gtaylor** [comment]
> But I may not want to max out my k8s cluster. I, for one, would like to put limits on how many resources my build process can use and queue up anything beyond that, without sucking up all the ram on my cluster.

@clintberry Then put resource limits on your namespace. Kubernetes offers all kinds of ways for you to throttle your builds.

**Comment by so0k** [comment]
>In a world where we have elastic compute and are pushing the idea of serverless, the build queue feels so 2003.

Gitlab CI with Multi runner and `docker-machine` provisioner, kind of does this (albeit in an ugly way)
https://github.com/honestbee/gitlab-infra#registering-multi-runner

@MacTynow might be interested as he set up our drone Helm charts

**Comment by ekozan** [comment]
i'm realy interessed about this one,

**Comment by bradrydzewski** [comment]
Another option worth considering is to create a version of the `drone agent` that only runs a single build, and is bundled with dind for running build steps. This would take advantage of the kubernetes scheduler without requiring any massive/fundamental changes to drone.

I realize this isn't the most optimal soution, but keep in mind that I am thinking of baby steps here. It is an interim solution ... you know ... _better a diamond with a flaw than a pebble without_ 

I also think the google folks might be doing something similar, but not sure
https://github.com/GoogleCloudPlatform/jenkernetes#note

__pros__ to this approach:

1. does not require a major overhaul to how drone works (major bonus here)
2. does not mutate the host
3. does not schedule containers on the k8s host that k8s is not aware of
3. let's us work with kubernetes in a more native way

__cons__ to this approach:

1. inability to distribute a single build across multiple machines
2. dind may not work in some environments (will docker ever support rootless containers?)
3. ephemeral build environments like this lose ability to cache docker image downloads. Mounting the cache is probably not an option since it is not safe for concurrent use by multiple daemons

__technical considerations:__

1. the drone-docker plugins uses dind. Can we run dind in dind?

And in case someone is considering posting a link to [this blog post](http://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/) keep in mind that it is quite old and docker has seen much progress since. I personally [support](https://discuss.drone.io/t/why-some-drone-plugins-run-docker-in-docker/209) using dind where it makes sense, and do not view it as an absolute evil :)

**Comment by tonglil** [comment]
> drone schedules special one-time agent to run on kubernetes cluster

1. To clarify, would this be a [Job resource](https://kubernetes.io/docs/user-guide/jobs/)? Because I think that really fits the bill here.

> inability to distribute a single build across multiple machines

1. Why/where is this a benefit? Sorry I didn't find anything concrete on this anywhere earlier on.

> dind may not work in some environments

and

> the drone-docker plugins uses dind. Can we run dind in dind?

2. I am not sure if you want to run dind completely, run drone-docker (and other whitelisted drone plugins) as a sibling and the build in dind, or run all the containers as siblings (probably not idea for the drone-docker plugin's isolation levels, though we have experienced the desire to be able to specify the use the caching of layers for large builds. One way maybe of "trusted repos"?).

3. On another note, I am not sure what the consequences of having Drone run "sibling" containers next to K8s managed containers are (K8s cleans up containers and images when disk space/inodes get low, but not sure if it would touch the ones Drone created).

**Comment by bradrydzewski** [comment]
> To clarify, would this be a Job resource? Because I think that really fits the bill here.

Yes probably, I am still learning this stuff 😄 

> Why/where is this a benefit? Sorry I didn't find anything concrete on this anywhere earlier on.

There is another github issue #1814 that discusses a fan-out to run build steps in parallel on multiple machines. This probably needs to remain out of scope for now.

> I am not sure if you want to run dind completely, run drone-docker (and other whitelisted drone plugins) as a sibling and the build in dind, or run all the containers as siblings

If you are running drone:0.5 agents on Kubernetes you are mounting the docker socket into the agent container, and the agent is launching sibling containers. The complaints I hear is that people don't like agents bypassing Kubernetes and interacting directly with the host machine docker daemon.

I think having some sort of roadmap would definitely help here, since I think this will take multiple iterations before we get to the ideal implementation. Should the initial implementation focus on using the Kubernetes scheduler and eliminating agents? If yes, the initial implementation should probably continue using sibling containers like you mentioned. If the initial focus should be on avoiding direct interaction with the host machine docker daemon, then we probably need to look at other solutions.

> On another note, I am not sure what the consequences of having Drone run "sibling" containers next to K8s managed containers are (K8s cleans up containers and images when disk space/inodes get low, but not sure if it would touch the ones Drone created).

This has not been raised as an issue, but perhaps the teams using drone:0.5 on Kubernetes can provide more insight.

**Comment by jmccann** [comment]
> the drone-docker plugins uses dind. Can we run dind in dind?

Yes you can.  And I was just playing with gitea/drone inside minikube and did such a setup with an agent

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: drone-agent
spec:
  replicas: 1
  template:
    metadata:
      labels:
        name: drone-agent
    spec:
      containers:
        - image: docker:1.12.6-dind
          name: dind
          ports:
            - containerPort: 2375
              protocol: TCP
          securityContext:
            privileged: true
        - image: drone/drone:0.5
          imagePullPolicy: Always
          name: drone-agent
          command:
            - "/drone"
            - "agent"
          env:
            - name: I_UNDERSTAND_I_AM_USING_AN_UNSTABLE_VERSION
              value: "true"
            - name: I_AGREE_TO_FIX_BUGS_AND_NOT_FILE_BUGS
              value: "true"

            - name: DRONE_DEBUG
              valueFrom:
                configMapKeyRef:
                  name: drone-config
                  key: agent.debug.is.enabled
            - name: DRONE_SERVER
              valueFrom:
                configMapKeyRef:
                  name: drone-config
                  key: agent.drone.server.url
            - name: DRONE_SECRET
              valueFrom:
                secretKeyRef:
                  name: drone-secrets
                  key: server.secret

            - name: DOCKER_HOST
              value: tcp://localhost:2375
          resources:
            requests:
              cpu: 1
              memory: 1G
```

And it seemed to work just fine with the following `.drone.yml`
```yaml
pipeline:
  build:
    image: alpine:3.4
    commands:
      - echo 'hi'

  docker:
    image: plugins/docker:latest
    repo: jmccann/test
    dry_run: true
```

Also, it seems pretty simple to launch/run/delete singular containers in kubernetes even:
```
$ kubectl run busybox --image=busybox sleep 5  --restart='Never'
pod "busybox" created
$ kubectl get pods --show-all
NAME                            READY     STATUS    RESTARTS   AGE
busybox                         1/1       Running   0          9m
$ kubectl get pods --show-all
NAME                            READY     STATUS    RESTARTS   AGE
busybox                         1/1       Running   0          9m
$ kubectl get pods --show-all
NAME                            READY     STATUS      RESTARTS   AGE
busybox                         0/1       Completed   0          9m
$ kubectl delete pods busybox
pod "busybox" deleted
$ kubectl get pods --show-all
NAME                            READY     STATUS    RESTARTS   AGE
```

Seems really similar to `docker run` command.

I'm working on cleaning up my gitea/drone minikube config with instructions to maybe help people get to playing with Drone in kubernetes locally.

**Comment by paultiplady** [comment]
Interested in following along with this too, I'm new to Drone but familiar with Kubernetes. @gtaylor you mentioned an early pod-based approach a while go in this thread, is that code available somewhere to look at?

**Comment by gtaylor** [comment]
@paultiplady That was me doing some tinkering outside of Drone with the Kubernetes API. I couldn't figure out anything that I thought was much of an improvement on the current state of things (running agents).

**Comment by ekozan** [comment]
I really love @jmccann solution need to be more tested :P

**Comment by derekperkins** [comment]
@bradrydzewski Did you and @kelseyhightower end up working on this at NEXT?

**Comment by bradrydzewski** [comment]
@derekperkins we did not, but I think I'm pretty clear on the design he proposed. I'm planning to take a shot at a basic implementation in the coming weeks. I'll post updates as I have them. Was good meeting you at Next!

**Comment by bradrydzewski** [comment]
This is a summary of the design he and I discussed:

1. create per-build namespace
2. create host-path volume
3. start a pod for each pipeline step, mounting the host-path volume
4. delete namespace

So this will be a native Kubernetes solution. Drone will not have any direct interaction with the host machine Docker daemon.

**Comment by ekozan** [comment]
Hello :) 

Just somme remarks 

1. Create Networkpolicy at the same time will be awesome :) https://kubernetes.io/docs/user-guide/networkpolicies/
2. Why not deploy an nfs server with drone for kuernetes ? :) so we are not node dependant 
https://github.com/kubernetes-incubator/external-storage/tree/master/nfs
And we can setup network policies for safety 
3. can we use jobs ? :)

**Comment by bradrydzewski** [comment]
Yes, we certainly should support nfs if available to the kubernetes environment. It would not be a requirement, though.

In terms of jobs, the were not recommend because they restart on non-zero exit code. So when a build step fails kubernetes will retry, regardless of restart policy.

**Comment by gtaylor** [comment]
Kubernetes on NFS is super shoddy and unreliable right now, FYI. I wouldn't suggest it for anything more than tinkering.

**Comment by ekozan** [comment]
I use it on my dev server without any problems 
but ok maybe more simple let's user decide ;) hostpath or volume with storage class

**Comment by webwurst** [comment]
Is there any early implementation to try out for some brave Kubernauts? :smile:

**Comment by automaticgiant** [comment]
so, nomad makes more sense to me here, though certainly it would make sense to have a drone plugin to deploy/upgrade/update a k8s job/service/whatever.
- 7 drivers (Docker, Isolated Fork/Exec, Java, LXC, Qemu, Raw Fork/Exec, Rkt)
- batch jobs - get queuing if you want, or scale up a nomad cluster on demand with terraform to get through them faster, or use contstraints for dedicated targets
- supports most popular arch/os combinations (darwin means you can ios, right?)
- good abstraction in `nomad agent -dev` (simultaneous client/server mode) for the various container systems anyway even if you don't go further than that.
- layer on a persistence thing if you want, could just constrain to same node and use shared local storage... plenty of options
- layer on other things optionally according to pipeline, like external secrets, service discovery, whatever
can just build it directly into the same binary as drone, or run in dev mode, or connect to a deployed cluster

thoughts?

**Comment by gtaylor** [comment]
@automaticgiant Kubernetes has pretty much all of those capabilities, with the added benefit of the majority mindshare in container orch. By contrast, I'm not sure that I've ever heard of someone using Nomad in production.

**Comment by automaticgiant** [comment]
several teams at capital one at least. (we might even be able to contribute to it.)
it would also synergize with some of our teams' production adoption of nomad.
i don't have anything against k8s per se, but one of the reasons i advocate for drone over jenkins is the flexibility of just hooking up whatever containers instead of having to deal with managing build environments on slaves. the next pain point to overcome, to me, is increase the flexibility of not having to use docker. (i see k8s supports rkt too, which is nice, but still not as flexible.)
my team had a use case where it was desired to build a docker image, then test it using inspec, from an inspec container. inspec supports talking to the docker daemon and running tests against a container without installing inspec in the container. in this case, we gave up because there didn't seem to be a practical way to get the stage container id that we built to provide to the inspec container. this kindof reflects on the contentious design of docker, having a daemon, instead of a container-per-process model, like rkt or lxc. and though k8s has rkt now, the point is not the specific technology, but having more tools to solve our problems.

any thoughts about how `drone exec` might work with k8s? it would be simple with nomad, but maybe simple, if more constrained, with minikube.

i also really like how easy it is to get started with drone - reminds me of nomad, though minikube is making strides. k8s seems too complicated for drone.
it may just be personal preference, but i think a more portable, flexible, and lightweight cluster scheduler that you can run natively instead of in a vm (my understanding of minikube) seems like a more appropriate target for drone, but then again, if the scheduler is modular, we can both have and eat our own favorite cake.

i'm doing alot of work around making devops simple and elegant, and alot of that basis is composing tools that do one thing well, and only when needed (ie don't add X if you don't need it). so while a k8s scheduler may make sense for you and work in your use case, nomad may work better for me. and they are both valid uses. i don't mean to tell you you are wrong for using k8s - you do you. contribute that. but from my perspective, if there were to be "team" (and by that, i mean brad...) work on a secondary executor system, nomad's operation/architecture/philosophy seems to be more closely aligned with drone's.
but again, like nomad, taking nomad's approach of multiple executor types might be what really sets drone apart from the rest.

**Comment by minhchuduc** [comment]
@automaticgiant thanks God, i found a person have the same idea with me :D I think Nomad more fit for Drone, too. But i not sure which way we should run Drone on Nomad (**at current time**). Which Nomad driver should we use to run Drone-agent? If we launch jobs in drone-agent with DinD style, which issues could happen?
@gtaylor i hear that http://imgix.com moved most of their auxiliary services to Nomad. 
imgix.com has nearly 1000 servers (VM) but not sure how many of it running Nomad. (They're using Consul for more than 2 years, it also a reason they choose Nomad)

**Comment by paultiplady** [comment]
@automaticgiant the scope of this issue is "Let's re-open this but with a slightly adjusted scope of adding experimental support for an alternate kubernetes engine, alongside the existing docker engine."

You should probably open a new issue for Nomad support; that will be a separate engine, and doesn't preclude a Kubernetes implementation. The discussion of Nomad will include a different set of people as well (I don't personally have any interest in following that discussion, but I think it's a great idea to investigate support for multiple orchestrators).

**Comment by bradrydzewski** [comment]
Yes, agreed with @paultiplady that we should track a separate issue. I was meaning to reply to some of these comments sooner, so I do want to quickly provided some thoughts ... I believe the runtime engines should be pluggable. I would therefore like to spend time making sure we have the correct abstractions in place to plugin alternate implementations, and allow community members to bring their own experimental runtimes.

In general the stance I take is that most features should start as external plugins before we consider inclusion in drone core. And in many cases features should stay plugins to keep drone core small and focused. But discussing what gets included or not is getting ahead of ourselves, because right now we are discussing vaporware :)

So I think the separate issue should focus on Nomad, and perhaps more broadly, on what sort of abstractions are required to make this capability pluggable.

**Comment by bradrydzewski** [comment]
OK, issue https://github.com/drone/drone/issues/2116 created for Nomad and pluggable runtime engines. Now back to kubernetes :)

**Comment by ekozan** [comment]
somebody work on it ?

**Comment by gtaylor** [comment]
We (at Reddit) may after Kubernetes 1.8. There are a few volume-related things that are still being hashed out that would make this much less clumsy. Right now I don't feel like there is a significant benefit to a Kubernetes-specific backend since agents work really well.

The biggest usage case we have for it is:

1. A build is triggered for a repo that requires elevated IAM privs
1. Instead of having an agent running 100% of the time (but almost always idle) with the specific elevated priv set, we'd fire off a Kubernetes Job with the IAM privs on-demand.

For us it's not a big deal for us to leave an elevated-priv agent running that doesn't see much activity, relative to the cost of implementing a Kubernetes backend.

Are there any other huge benefits to starting work on this now, or is it mostly for sexiness?

**Comment by sethpollack** [comment]
@gtaylor how do you configure your agents? Do you run dind containers on the agent deployments or do you have the agents use the hosts docker socket?

**Comment by gtaylor** [comment]
@sethpollack We do somewhat like the WIP drone chart in kubernetes/charts, in that we run a Pod with an agent and a DinD container (though you could easily run more than one agent container per pod). We didn't want to use the host's docker socket, since that would circumvent resource limits/requests and represent more risk of breakouts. Also, we wanted to use a more recent version of Docker than the now-long-in-tooth 1.12 that is certified for Kubernetes 1.7.

In our specific case, we want to handle image and container pruning outside of the Docker agent, so we've got a PR to make these configurable flags: https://github.com/drone-plugins/drone-docker/pull/142

We'll probably have a CronJob or a sidecar container manage pruning for the pod. Running more than one agent per-pod would, in theory, increase your hit rate at the cost of making your pods larger units to be scheduled.

**Comment by sethpollack** [comment]
Yeah, I'm doing something similar. So another use case for using a k8s backend would be eliminating the need for dind.

**Comment by gtaylor** [comment]
FWIW, I'd _want_ DinD due to Kubernetes being hyper-conservative with Docker versions, plus the extra line of defense.

**Comment by sethpollack** [comment]
So how would that work with a k8s backend?

**Comment by gtaylor** [comment]
It depends on how you define "k8s backend". For us, that'd mean firing off Kubernetes Jobs for individual builds instead of an always-on agent (that is always using cluster resources even in an idle state).

The nature of the Pod is an implementation detail, as far as I'm concerned. The scheduling is the main event.

**Comment by sethpollack** [comment]
Yeah, but how would you deal with layer caching etc. if you are not using the hosts docker socket and the agents are one-off runs?

**Comment by gtaylor** [comment]
Our approach is outlined on https://github.com/drone/drone/issues/1815#issuecomment-326021345

We get layer caching, isolation from our host's Docker daemon, a newer version of Docker than our host's Docker daemon (1.12.x), and we don't circumvent resource tracking (as is the case with mounting the host's Docker daemon).

**Comment by gtaylor** [comment]
Dump your DinD container out into a Service, do basically the same thing.

**Comment by ekozan** [comment]
I go to DInd 
For info my helm charts -> http://git.eko.ovh/ekozan/helm-charts/src/drone/incubator/drone/templates

**Comment by LFlare** [comment]
Sorry to cut-in but my development workflow has recently been prioritizing budget. My current setup involves having a single RPI3B serving as my central server including hosting Gitea and Drone server.

However, just awhile ago, I came across [hyper.sh](https://hyper.sh) and I felt that the key concepts behind it seems so much in-tune with what a CI agent needs. In particular, the build server doesn't have to run 24/7 _hyper.sh bills per second_ and it's more or less based around Docker as a whole.

Unfortunately, as of this writing, it seems drone agents are fairly limited in capabilities? Is integration with something like _hyper.sh_ feasible?

**Comment by gtaylor** [comment]
@LFlare can we take hyper.sh discussion to another issue or better yet, discourse? This issue is already getting kind of out of hand with off-topic stuff.

**Comment by LFlare** [comment]
@gtaylor I unfortunately do not intend to discuss this in a thread-like manner akin to that of Discourse. Never mind that, I don't think @bradrydzewski has the time for this anyways.

**Comment by bradrydzewski** [comment]
> Unfortunately, as of this writing, it seems drone agents are fairly limited in capabilities? Is integration with something like hyper.sh feasible?

We can plugin different container backends (kubernetes, hyper, etc) as long as they can implement the below interface. I personally do not see an issue with drone being flexible enough. If anything, I would question whether or not hyper will be flexible enough.

```Go
// Engine defines a container orchestration backend and is used
// to create and manage container resources.
type Engine interface {
	// Setup the pipeline environment.
	Setup(*Config) error
	// Start the pipeline step.
	Exec(*Step) error
	// Kill the pipeline step.
	Kill(*Step) error
	// Wait for the pipeline step to complete and returns
	// the completion results.
	Wait(*Step) (*State, error)
	// Tail the pipeline step logs.
	Tail(*Step) (io.ReadCloser, error)
	// Destroy the pipeline environment.
	Destroy(*Config) error
}
```

I think the real issue here is not technical. The problem is despite clear interest in additional container runtimes and backends we do not have volunteers to implement or alternatively sponsor features (financially).

I agree with Greg that we should track hyper as a separate issue.

I also think we probably should lock this issue pending a volunteer or sponsor. If anyone is reading this thread and interested in implementing native kubernetes or hyper support, or willing to fund feature development, please open a thread at https://discourse.drone.io/

I will unlock this issue once we have progress to report.

**Comment by bradrydzewski** [comment]
unlocking this issue now that we have a volunteer and a base implementation in place. If you have any questions please direct to @metalmatze. The kubernetes runtime source code is available here https://github.com/drone/drone-kubernetes-runtime

@metalmatze gave a demo of the kubernetes runtime at container days. I will post a link here to the video when one is available:
https://twitter.com/thenewstack/status/1009059265032671232

note that the kubernetes runtime will require drone version 0.9 which is not yet released. I do not have an estimate release date to share at this time. In the mean time, you can build, test and run the kubernetes runtime from the command line if you want to take a look or get your hands dirty.

**Comment by combor** [comment]
@bradrydzewski is this included in upcoming 1.0?

**Comment by zetaab** [comment]
@combor drone-kubernetes-runtime repo says WIP. Anyways in drone-k8s-engine it seems that there is requirement for RWX persistent volumes, at least we do not have those in our environment.

**Comment by tboerger** [comment]
The kubernetes runtime repo had been the initial development, AFAIK it's moved into the drone-runtime repo part of a different development branch: https://github.com/drone/drone-runtime/tree/kubernetes

**Comment by bradrydzewski** [comment]
I am very excited to CLOSE this issue as COMPLETE
https://blog.drone.io/drone-goes-kubernetes-native/

**Comment by JeanMertz** [comment]
I haven't responded here in a while, but I wanted to thank you and everyone else for the hard work on making this happen @bradrydzewski ❤️ 

The combination of Drone and Kubernetes (especially when using a hosted solution such as GKE), makes a self-hosted CI/CD setup simpler, yet more powerful than ever. Having Drone be open-source and as extensible as it is, is icing on the cake, as it'll allow us to make Drone work just the way we need it to.

I'm looking forward to start experimenting with this. Awesome job.

**Comment by bradrydzewski** [comment]
@JeanMertz thank you so much for the kind words. I think there is still a ton of room for improvement and I am really excited for what is next (not sure what that is, but I'm excited for it). Also huge thanks for @metalmatze. He started working on this in October 2017, so I really appreciate his help (and patience) getting everything rolled out.

**Comment by zetaab** [comment]
This is cool thing! However, at least we would like to see builds without docker agent. The problem with docker is that we need to mount docker sockets inside container, which is not really secure. I am thinking could we somehow use kaniko instead? https://github.com/GoogleContainerTools/kaniko

**Comment by JeanMertz** [comment]
> However, at least we would like to see builds without docker agent

Wait... unless I'm misreading, I believe _this is without mounting Docker sockets_. The blog mentions a job is started as a _native_ Kubernetes job, which in turn spins up _native_ Kubernetes pods and "secrets" to share any configuration between the jobs.

The "old" way of doing Drone on Kubernetes was indeed with mounted Docker sockets (with all the security risk and lack of maintainability included), but this announcement removes all of that, and you no longer have to concern yourself with the container engine running below Kubernetes.

**Comment by zetaab** [comment]
@JeanMertz WOW if that is true! need to test this

**Comment by JeanMertz** [comment]
> @JeanMertz WOW if that is true! need to test this

It is! Check out this [work in progress guide](https://discourse.drone.io/t/contributing-to-drone-for-kubernetes/3159):

> The Drone Kubernetes Runtime takes a .drone.yml configuration and translates it into Kubernetes native pods, secrets, and services. So what does this mean if you are already running Drone on Kubernetes today? It means no more build agents. No more mouting the host machine Docker socket or running Docker in Docker. If you are into buzzwords, it means Drone is now Kubernetes Native.

**Comment by bradrydzewski** [comment]
yep, this is correct. The Pipeline steps are launched as Kubernetes Pods. There are no more agents, and no more mounting Docker sockets :)

**Comment by malcolmholmes** [comment]
@bradrydzewski any pointers to how to start this (as opposed to Docker based Drone)? This is very exciting as I was just starting to battle with Drone in Kubernetes. Sample k8s yaml manifests, or other startup hints would be extremely useful. Thanks!

**Comment by bradrydzewski** [comment]
@malcolmholmes there is work being done on an updated helm chart that might be helpful. See https://github.com/helm/charts/pull/9617

There are also some _very high level_ docs [here](https://docs.drone.io/installation/github/kubernetes/), but they only really show you the absolute basics and would require you to setup a more complete configuration.

**Comment by malcolmholmes** [comment]
@bradrydzewski thanks for those. Very helpful.

**Comment by malcolmholmes** [comment]
Hi, I have followed the instructions as best I can. I am using gitlab.com, and get this:

`{"error":"Source code management system not configured","level":"fatal","msg":"main: source code management system is not configured","time":"2018-12-16T22:40:53Z"}`. Would you expect this version to work with Gitlab?

Here's the container config:

```
    spec:
      containers:
      - env:
        - name: DRONE_GITLAB
          value: "true"
        - name: DRONE_ADMIN
          value: *A_VALUE_HERE*
        - name: DRONE_LOGS_DEBUG
          value: "true"
        - name: DRONE_KUBERNETES_ENABLED
          value: "true"
        - name: DRONE_KUBERNETES_NAMESPACE
          value: build-jobs
        - name: DRONE_GITLAB_CLIENT
          valueFrom:
            secretKeyRef:
              key: drone_gitlab_client
              name: drone
        - name: DRONE_GITLAB_SECRET
          valueFrom:
            secretKeyRef:
              key: drone_gitlab_secret
              name: drone
        - name: DRONE_RPC_SECRET
          valueFrom:
            secretKeyRef:
              key: drone_shared_secret
              name: drone
        image: drone/drone:1.0.0-rc.2
        name: drone
        ports:
        - containerPort: 80
        - containerPort: 443
```

**Comment by bradrydzewski** [comment]
@malcolmholmes many of those environment variable names are incorrect. I also recommend creating thread in our mailing list as opposed to seeking support through this existing issue. https://discourse.drone.io

```diff
      - env:
-       - name: DRONE_GITLAB
-         value: "true"
        - name: DRONE_ADMIN
          value: *A_VALUE_HERE*
        - name: DRONE_LOGS_DEBUG
          value: "true"
        - name: DRONE_KUBERNETES_ENABLED
          value: "true"
        - name: DRONE_KUBERNETES_NAMESPACE
          value: build-jobs
-       - name: DRONE_GITLAB_CLIENT
+       - name: DRONE_GITLAB_CLIENT_ID
          valueFrom:
            secretKeyRef:
              key: drone_gitlab_client
              name: drone
-       - name: DRONE_GITLAB_SECRET
+       - name: DRONE_GITLAB_CLIENT_SECRET
          valueFrom:
            secretKeyRef:
              key: drone_gitlab_secret
              name: drone
        - name: DRONE_RPC_SECRET
          valueFrom:
            secretKeyRef:
              key: drone_shared_secret
              name: drone
```
